src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -64,54 +67,43 @@ class CassandraJournal extends AsyncWriteJournal with CassandraRecovery with Cas
+      .zip(flattenedMessages)
+      .map(aw => Try {
+        Serialized(aw._1, aw._2.sequenceNr, aw._2.persistenceId, persistentToByteBuffer(aw._2))
       })
     journalSequenceNr = newJournalSequenceNr
 
-    val result = serialized.map(a => a.map(_ => ()))
-
-    val byPersistenceId = serialized.collect({ case Success(caw) => caw }).groupBy(_.persistenceId).values
-    val boundStatements = byPersistenceId.map(statementGroup)
-
-    val batchStatements = boundStatements.map({ unit =>
-      executeBatch(batch => unit.foreach(batch.add))
-    })
-    val promise = Promise[Seq[Try[Unit]]]()
-
-    Future.sequence(batchStatements).onComplete {
-      case Success(_) => promise.complete(Success(result))
-      case Failure(e) => promise.failure(e)
-    }
-
-    promise.future
-  }
-
-  private def statementGroup(atomicWrites: Seq[SerializedAtomicWrite]): Seq[BoundStatement] = {
-    val firstJournalSequenceNr = atomicWrites.head.payload.head.journalSequenceNr
-    val lastJournalSequenceNr = atomicWrites.last.payload.last.journalSequenceNr
-
-    val maxPnr = partitionNr(firstJournalSequenceNr)
-    val minPnr = partitionNr(lastJournalSequenceNr)
-    val persistenceId: String = atomicWrites.head.persistenceId
-    val all = atomicWrites.flatMap(_.payload)
-
-    // reading assumes sequence numbers are in the right partition or partition + 1
-    // even if we did allow this it would perform terribly as large C* batches are not good
-    require(maxPnr - minPnr <= 1, "Do not support AtomicWrites that span 3 partitions. Keep AtomicWrites <= max partition size.")
-
-    val writes: Seq[BoundStatement] = all.map { m =>
-      preparedWriteMessage.bind(journalId, maxPnr: JLong, m.journalSequenceNr: JLong, persistenceId, m.sequenceNr: JLong, m.serialized)
-    }
-    // in case we skip an entire partition we want to make sure the empty partition has in in-use flag so scans
-    // keep going when they encounter it
-    if (partitionNew(firstJournalSequenceNr) && minPnr != maxPnr) writes :+ preparedWriteInUse.bind(journalId, minPnr: JLong)
-    else writes
+    val byPersistenceId = serialized
+      .collect({ case Success(caw) => caw })
+
+    val boundJournalEntries: (String, scala.Seq[Serialized]) => scala.Seq[BoundStatement] =
+      (persistenceId, entries) => {
+
+        val maxPnr = partitionNr(entries.last.journalSequenceNr)
+
+        entries.map{ e =>
+          preparedWriteMessage.bind(
+            journalId,
+            maxPnr: JLong,
+            e.journalSequenceNr: JLong,
+            e.persistenceId,
+            e.sequenceNr: JLong,
+            e.serialized)
+        }
+      }
+
+    writeBatch[String, Serialized](
+      Map(journalId -> byPersistenceId),
+      boundJournalEntries,
+      _.journalSequenceNr,
+      (persistenceId, minPnr) => preparedWriteInUse.bind(journalId, minPnr: JLong))
+      .map(_.to[scala.collection.immutable.Seq])
```

Comment:
at some point we should have an optimized path for this when `messages` contains one `AtomicWrite`, i.e. the normal case for `persist`